timeout_handler.go

```diff
@@ -0,0 +1,135 @@
+package simplex
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+)
+
+type ID string
+
+type TimeoutTask struct {
+	ID      ID
+	Task    func()
+	Timeout time.Time
+
+	index int // for heap to work more efficiently
+}
+
+func NewTimeoutTask(ID ID, task func(), timeout time.Time) *TimeoutTask {
+	return &TimeoutTask{
+		ID:      ID,
+		Task:    task,
+		Timeout: timeout,
+	}
+}
+
+type TimeoutHandler struct {
+	lock sync.Mutex
+
+	tasks map[ID]*TimeoutTask
+	heap  TaskHeap
+	now   time.Time
+}
+
+func NewTimeoutHandler(startTime time.Time) *TimeoutHandler {
+	return &TimeoutHandler{
+		now:   startTime,
+		tasks: make(map[ID]*TimeoutTask),
+		heap:  TaskHeap{},
+	}
+}
+
+func (t *TimeoutHandler) GetTime() time.Time {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.now
+}
+
+func (t *TimeoutHandler) Tick(now time.Time) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	// update the time of the handler
+	t.now = now
+
+	// go through the heap executing relevant tasks
+	for t.heap.Len() > 0 {
+		next := t.heap[0]
+		if next.Timeout.After(t.now) {
+			break
+		}
+
+		heap.Pop(&t.heap)
+		delete(t.tasks, next.ID)
+		go next.Task()
```

Comment:
I would rather we not create goroutines in a non-controlled manner like that. Can we not run this in this current goroutine?